src/store/helpers.js

```diff
@@ -26,3 +28,173 @@ export function onlyHandleAPIError (error, handleFn) {
+      byId: {},
+      byAction: {},
+    },
+    getters: {
+      status: state => (actionName, id) => {
+        if (id) {
+          return { ...defaultStatus, ...(state.byId[id] && state.byId[id][actionName]) }
+        }
+        else {
+          return { ...defaultStatus, ...state.byAction[actionName] }
+        }
+      },
+    },
+    actions: {
+      clear ({ commit }, [actionName, id]) {
+        commit('clear', { actionName, id })
+      },
+    },
+    mutations: {
+      update (state, { actionName, id, value }) {
+        if (id) {
+          if (!state.byId[id]) Vue.set(state.byId, id, {})
+          Vue.set(state.byId[id], actionName, value)
+        }
+        else {
+          Vue.set(state.byAction, actionName, value)
+        }
+      },
+      clear (state, { actionName, id }) {
+        if (id) {
+          if (state.byId[id]) {
+            Vue.delete(state.byId[id], actionName)
+            if (Object.keys(state.byId[id]).length === 0) {
+              Vue.delete(state.byId, id)
+            }
+          }
+        }
+        else {
+          Vue.delete(state.byAction, actionName)
+        }
+      },
+    },
+  }
+}
+
+export function withMeta (actions, { namespace = 'meta', idPrefix = '', findId = defaultFindId } = {}) {
+  const wrappedActions = {}
+  for (const [actionName, action] of Object.entries(actions)) {
+    wrappedActions[actionName] = wrapAction({ namespace, actionName, action, findId })
+  }
+  return wrappedActions
+}
+
+export function withPrefixedIdMeta (idPrefix, actions, options) {
+  return withMeta(actions, { ...options, idPrefix })
+}
+
+function defaultFindId (data) {
+  if (typeof data === 'number') {
+    return data
+  }
+  else if (typeof data === 'object' && data.hasOwnProperty('id')) {
+    return data.id
+  }
+}
+
+function wrapAction ({ namespace, actionName, action, idPrefix, findId }) {
+  return async function (ctx, data) {
+    const { commit, getters } = ctx
+    let id = findId(data)
+
+    if (id && idPrefix) id = idPrefix + id
+
+    if (getters[`${namespace}/status`](actionName, id).pending) {
+      throw new Error(`action already pending for ${actionName}/${id}`)
+    }
+
+    const runAction = () => action.apply(this, arguments)
+    const update = value => commit(`${namespace}/update`, { actionName, id, value })
+    const clear = () => commit(`${namespace}/clear`, { actionName, id })
+
+    update({ pending: true })
+
+    try {
+      await runAction()
+      clear()
+      return true
+    }
+    catch (error) {
+      if (isValidationError(error)) {
+        update({ validationErrors: error.response.data })
+        return false
+      }
+      else {
+        // some other error, can't handle it here
+        clear()
+        throw error
+      }
+    }
+  }
+}
+
+export function metaStatusesWithId (getters, actions, id) {
+  const result = {}
+  for (let action of actions) {
+    result[action + 'Status'] = getters['meta/status'](action, id)
+  }
+  return result
+}
+
```

Comment:
Building identifiers via string replacement makes "grepping" around in the codebase harder. I think it could be worthwhile in this case, although I would be cautious to continue metaprogramming like this.

Reply:
I agree. Hopefully this metaprogramming for metadata is perfect for this usecase, so it hides away the crap, but the main actual data can just still use normal actions/mutations.